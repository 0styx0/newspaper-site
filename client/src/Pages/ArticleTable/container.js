import React from 'react';
import fetchFromApi from '../../helpers/fetchFromApi';

import ArticleTable from './';

export default class ArticleTableContainer extends React.Component {

    constructor() {
        super();

        this.putData = this.putData.bind(this);

        this.state = {
            history: []
        };
    }

    async putData(num = null) {

        const lastPath = +window.location.pathname.split("/").slice(-1);

        num = (isNaN(lastPath) || ((!isNaN(num)) && num)) ? num : lastPath;


        // use cached results (don't load new info)
        if (this.state.history[num]) {
            this.setState({
                articles: this.state.history[num].articles,
                issueInfo: this.state.history[num].issueInfo
            });

            window.history.pushState({}, `Issue ${num}`, num);
        }
        else {

            const data = await this.getData(num);
            const json = await data.json();

            window.history.pushState({}, `Issue ${num}`, isNaN(lastPath) ? "modifyArticles/"+json[2].num : json[2].num);

            this.setArticleInfo(json);
        }
    }

    componentWillMount() {

        this.putData(null);
    }

    async getData(num) {

         return await fetchFromApi(`articleGroup?articlesFor=${num}`);

    }


    setArticleInfo(data) {

        const history = [...this.state.history];
        history[data[2].num] = {articles: data[0], issueInfo: data[2]};

        this.setState({
            articles: data[0].length ? data[0] :  null,
            issueInfo: data[2],
            history
        });
    }

    render() {

            return <ArticleTable
                     articles={this.state.articles}
                     issue={this.state.issueInfo}
                     update={(e) => this.putData(e.target.value)}
                   />
    }
}
