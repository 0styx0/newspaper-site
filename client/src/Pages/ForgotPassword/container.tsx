import * as React from 'react';
import { graphql, withApollo } from 'react-apollo';
import { RecoverPassword } from '../../graphql/user';
import ForgotPassword from './';

interface Props {
    recoverPassword:
      (params: {
          query: typeof RecoverPassword,
          variables: { username: string, authCode: string, email: string }
        }
      ) => Promise<{message: string}>;
}

export class ForgotPasswordContainer extends React.Component<Props, {}> {

    constructor(props: Props) {
        super(props);

        this.onSubmit = this.onSubmit.bind(this);

    }

    /**
     * Sends data to server
     */
    onSubmit(target: HTMLFormElement) {

        this.props.recoverPassword({
            query: RecoverPassword,
            variables: {
                authCode: (target.querySelector('input[name=lastAuth]') as HTMLInputElement).value,
                username: (target.querySelector('input[name=username]') as HTMLInputElement).value,
                email: (target.querySelector('input[name=email]') as HTMLInputElement).value,
            }
        });
    }

    render() {
        return <ForgotPassword onSubmit={this.onSubmit} />;
    }
}

const ForgotPasswordContainerWithData =
    graphql(RecoverPassword, {name: 'recoverPassword'})(ForgotPasswordContainer as any);

export default withApollo(ForgotPasswordContainerWithData);
