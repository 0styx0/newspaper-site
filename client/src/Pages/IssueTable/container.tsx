import * as React from 'react';
import { compose, graphql } from 'react-apollo';
import { IssueQuery, IssueUpdate } from '../../graphql/issues';
import { Issue } from './interface.shared';
import IssueTable from './';

interface State {
    privateIssue: { // admins can change these (until public is true)
        public?: boolean;
        name?: string;
    };
    loaded: boolean;
}

interface Props {
    data: {
        loading: boolean;
        issues: Issue[]
    };
    mutate: Function;
}

/**
 * Creates a table of issues, in descending order with info of
 * number, name, datePublished, and views (sum of views of all articles in the issue)
 *
 * If user is admin (lvl 3) and issue is not public, user can change issue name and make it public.
 * In that case, @see IssueTable.allowEditsOfLastIssue for difference in UI
 */
export class IssueTableContainer extends React.Component<Props, State> {

    constructor() {
        super();

        this.onChangeIssueInfo = this.onChangeIssueInfo.bind(this);
        this.onSubmit = this.onSubmit.bind(this);

        this.state = {
            privateIssue: {},
            loaded: false
        };
    }

    /**
     * When graphql gets data from @see graphql/issues,
     * this turns it into 2d array of what will be table rows and saves it to state.issueInfo
     *
     * If current user is an admin, enables editing of most recent, unpublished issue @see this.allowEditsOfLastIssue
     */
    componentWillReceiveProps(props: Props) {

        if (!props.data.issues || this.props.data.issues.length > 0) {
            return;
        }

        this.setState({loaded: true});
    }

    /**
     * @param e {HTMLInputElement event}
     *
     * Adds input's value to this.state.privateIssue[input.name]
     *
     * @example if event.target = <input name="public" value="1" />, then after this, this.state.privateIssue.public = 1
     */
    onChangeIssueInfo(e: Event) {

        const target = e.target as HTMLInputElement;

        const privateIssue = Object.assign({}, this.state.privateIssue);
        privateIssue[target.name] = target.value;

        this.setState({
            privateIssue
        });
    }

    /**
     * @param e {HTMLFormElement event}
     *
     * Sends this.state.privateIssue data to server (so name and/or public status can be saved to db)
     */
    onSubmit(e: Event) {

        e.stopPropagation();
        e.preventDefault();

        this.props.mutate({
            variables: {
                name: this.state.privateIssue.name,
                public: !!this.state.privateIssue.public
            }
        });
    }

    render() {

        if (this.props.data.issues.length < 1) {
            return null;
        }

        return (
            <IssueTable
              key={+this.state.loaded}
              onSubmit={this.onSubmit}
              onChangeIssueInfo={this.onChangeIssueInfo}
              issues={this.props.data.issues}
            />
        );
    }
}


const IssueTableContainerWithData = compose(
    graphql(IssueQuery),
    graphql(IssueUpdate)
)(IssueTable as any);

export default IssueTableContainerWithData;