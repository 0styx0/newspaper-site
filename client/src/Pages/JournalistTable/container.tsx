import * as React from 'react';
import twoDimensionalSorter from '../../helpers/twoDimensionalSorter';
import { compose, graphql } from 'react-apollo';
import { UserQuery, UserUpdate, UserDelete } from '../../graphql/users';
import { User } from './interface.shared';
import JournalistTable from './';
import graphqlErrorNotifier from '../../helpers/graphqlErrorNotifier';

import './index.css';
import { ChangeEvent } from 'react';

export interface State {
    users: User[];
    usersToDelete: Set<string | undefined>;
    idLevelMap: Map<string, number>;
}

export interface Props {
    data: {
        loading: boolean;
        users: User[];
    };
    userUpdate: Function;
    userDelete: Function;
}

/**
 * Creates a table with info about users in it
 *
 * Anyone can see users' name (along with link to profile), number of articles, and total views that person has
 *
 * Logged in users can also see users' levels
 *
 * Logged in users with a level greater than one in the table can also modify that user's level and delete account
 */
export class JournalistTableContainer extends React.Component<Props, State> {

    constructor(props: Props) {
        super(props);

        this.sortInfo = this.sortInfo.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.onDelete = this.onDelete.bind(this);
        this.onIdLevelMap = this.onIdLevelMap.bind(this);

        this.state = {
            users: [],
            usersToDelete: new Set(),
            idLevelMap: new Map()
        };
    }

    /**
     * If data is loaded, sets state.users to props.data.users
     */
    componentWillReceiveProps(props: Props) {

        if (!props.data.users || this.state.users.length > 0) {
            return;
        }

        this.setState({
            users: props.data.users
        });
    }

    /**
     * Sorts by target.value
     */
    sortInfo(event: ChangeEvent<HTMLSelectElement>) {

        const sortBy = (event.target as HTMLSelectElement).value;
        const sortedData = twoDimensionalSorter(this.props.data.users.slice(), sortBy) as User[];

        this.setState({
            users: sortedData
        });
    }

    /**
     * @param e - {htmlInputEvent} whose target.value is a number
     * @param id - id of user
     *
     * Adds an mapping of id -> e.target.value to this.state.idLevelMap
     */
    onIdLevelMap(e: ChangeEvent<HTMLSelectElement>, id: string) {

        const target = e.target as HTMLSelectElement;
        const mapCopy = new Map(this.state.idLevelMap);

        mapCopy.set(id, +target.value);

        this.setState({
            idLevelMap: mapCopy
        });
    }

    /**
     * Adds input value to {@see this.state.usersToDelete} if input is checked, else removes the value
     *
     * @param {htmlInputEvent} - event from html checkbox
     */
    onDelete(e: ChangeEvent<HTMLInputElement>) {

        const target = e.target as HTMLInputElement;

        const copySet = new Set(this.state.usersToDelete);

        target.checked ? copySet.add(target.value) : copySet.delete(target.value);

        this.setState({
            usersToDelete: copySet
        });
    }

    /**
     * @param idLevelMap - map of id to level
     *
     * @return object where level => [ids that mapped to that level]
     *
     * @example convertMapToArrayOfJSON(new Map([['123', 1], ['456', 2], ['789', 1])) =>
     * [
     *   {
     *     level: 1
     *     ids: ['123', '789']
     *   },
     *   {
     *     level: 2,
     *     ids: ['456']
     *   }
     * ]
     */
    convertMapToArrayOfJSON(idLevelMap: Map<string, number>) {

        const data: {level: number; ids: string[]}[] = [];

        Array.from(idLevelMap).forEach(mapping => {

            const [id, level] = mapping;

            const indexOfLevel = data.findIndex(elt => elt.level === level);

            if (indexOfLevel !== -1) {

                data[indexOfLevel].ids.push(id);
            } else {

                data[data.length] = {
                    level,
                    ids: [id]
                };
            }
        });

        return data;
    }

    /**
     * Sends data to server, if there's data to be sent from this.state usersToDelete and/or idLevelMap
     */
    async onSubmit(target: HTMLFormElement) {

        const data = this.convertMapToArrayOfJSON(this.state.idLevelMap);

        const password = (target.querySelector('[name=password]') as HTMLInputElement).value;

        if (data.length > 0) {

            await graphqlErrorNotifier(
                this.props.userUpdate, {
                    variables: {
                        data,
                        password
                    }
                },
                'userUpdate'
            );

            // this line will only fire if sucess since graphqlErrorNotifier throws error otherwise
            const updatedUsers = [...this.state.users].map(user => {

                const updatedLevel = this.state.idLevelMap.get(user.id);
                return Object.assign(JSON.parse(JSON.stringify(user)), {
                    level: updatedLevel || user.level
                });
            });

            this.setState({
                users: updatedUsers,
                idLevelMap: new Map()
            });
        }

        if (this.state.usersToDelete.size > 0) {

            await graphqlErrorNotifier(
                this.props.userDelete, {
                    variables: {
                        ids: [...this.state.usersToDelete],
                        password
                    }
                },
                'userDeleted'
            );

            // this line will only fire if sucess since graphqlErrorNotifier throws error otherwise
            const existingUsers = this.state.users.filter(user => !this.state.usersToDelete.has(user.id));

            this.setState({
                users: existingUsers,
                usersToDelete: new Set()
            });
        }
    }

    render() {

        if (!this.state.users.length) {
            return null;
        }

        return (
            <JournalistTable
              key={this.state.users.map((user: User) => user.id).join(',')}
              users={this.state.users}
              onDelete={this.onDelete}
              onSortInfo={this.sortInfo}
              onSubmit={this.onSubmit}
              onLevelChange={this.onIdLevelMap}
            />
        );
    }
}

const JournalistTableContainerWithData = compose(
    graphql(UserQuery),
    graphql(UserUpdate, {name: 'userUpdate'}),
    graphql(UserDelete, {name: 'userDelete'})
)(JournalistTableContainer);

export default JournalistTableContainerWithData;
