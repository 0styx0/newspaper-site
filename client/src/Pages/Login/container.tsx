import * as React from 'react';
import { setJWT, getJWT } from '../../helpers/jwt';
import LoginForm from './';
import { UserLogin } from '../../graphql/user';
import { graphql, withApollo } from 'react-apollo';
import graphqlErrorNotifier from '../../helpers/graphqlErrorNotifier';

import './index.css';

interface Props {
    history: string[];
    loginUser: ( params:
    {
        variables: {
            username: string,
            password: string
        }
    } ) => Promise<{ data: { login: { jwt: string } }, errors?: {message: string}[] }>;
}

export class LoginFormContainer extends React.Component<Props, {}> {

    constructor() {
        super();

        this.setJWT = this.setJWT.bind(this);
    }

    /**
     * Gets jwt from server and saves to localStorage if valid login info is in form given
     *
     * @param e - even from form with [name=username], [name=password]
     */
    async setJWT(target: HTMLFormElement) {

        const username = (target.querySelector('[name=username]') as HTMLInputElement).value;
        const password = (target.querySelector('[name=password]') as HTMLInputElement).value;

        const { data } = await graphqlErrorNotifier(this.props.loginUser, {
            variables: {
                username,
                password
            }
        });

        setJWT(data.login.jwt);
        const jwt = getJWT();

        if (jwt.level) {

            this.props.history.push('/publish');
        }
        else if (jwt.id) {

            this.props.history.push('/authLogin');
        }
    }

    render() {

        return (
            <LoginForm onLogin={this.setJWT} />
        );
    }
}

const LoginFormContainerWithData = graphql(UserLogin, { name: 'loginUser' })(LoginFormContainer as any);

export default withApollo(LoginFormContainerWithData);
