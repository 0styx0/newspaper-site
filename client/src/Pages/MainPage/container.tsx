import * as React from 'react';
import { Article, Issue } from './shared.interfaces';
import MainPage from './';
import { graphql, compose, withApollo } from 'react-apollo';
import { ArticlePreviewIssueQuery, ArticlePreviewTagQuery } from '../../graphql/articles';
import { IssueInfoQuery } from '../../graphql/issues';

interface Props {
    client: {
        query: Function; // NOTE: the commented out stuff below IS CORRECT but typescript throws error. Not sure why
        //   (( params: { query: typeof ArticlePreviewQuery, variables: { issue?: number; tag?: string} } ) =>
        //     Promise<{ data: Article[] }>)
        //     |
        //   (( params: { query: typeof IssueInfoQuery, variables: { num?: number; } } ) =>
        //     Promise<{ data: [Issue] }>);
    };
    history: {listen: Function};
}

interface State {
    articles: Article[];
    issue: Issue;
}

export class MainPageContainer extends React.Component<Props, State> {

    unlisten: Function;

    constructor(props: Props) {
        super(props);

        this.state = {
           articles: [] as Article[],
            // should only use actual issue stuff (and not articles, rely on state.articles for that)
           issue: {} as Issue
        };
    }

    /**
     * Loads initial data and sets this.unlisten
     */
    async componentWillMount() {

        this.fetchInfo();

        this.unlisten = this.props.history.listen(() => {
            this.fetchInfo();
        });
    }

    componentWillUnmount() {

        this.unlisten();
    }

    /**
     * calls #fetchIssue and #fetchArticles
     */
    fetchInfo() {

        const requested = this.getRequested();

        const issue = Number.isNaN(+requested) ? 0 : +requested;

        // fixes bug where call setState when component is unmounted (due to history.listen)
        if (/^\/($)|(tag\/)|(issue\/\d+?$)/i.test(window.location.pathname)) {
            this.fetchArticles(issue, requested);
            this.fetchIssue(issue);
        }
    }

    /**
     * Get articles from server with issue or tag given by params
     */
    fetchArticles(issue: number, tag: string) {

        let articleQuery: Promise<{ data: { articles: Article[] } }>;

        if (tag === 'Current%20Issue' || !tag || (issue !== 0 && !isNaN(+issue))) {

            articleQuery = this.props.client.query({
                query: ArticlePreviewIssueQuery,
                variables: {
                    issue
                }
            }).then(({ data }: { data: { issues: [Article[]] } }) => ({data: data.issues[0] }));

        } else {

            articleQuery = this.props.client.query({
                query: ArticlePreviewTagQuery,
                variables: {
                    issue,
                    tag
                }
            });
        }

        articleQuery.then(({ data }: { data: { articles: Article[] } }) => {

            this.setState({
                articles: data.articles
            });
        });
    }

    /**
     * Get issue info from server
     */
    fetchIssue(issue?: number) {

        this.props.client.query({
            query: IssueInfoQuery,
            variables: {
                num: issue
            }
        }).then(({ data }: { data: { issues: [Issue] } }) => {

            this.setState({
                issue: data.issues[0]
            });
        });

    }

    /**
     * Gets the tag or issue that user wants to view
     */
    getRequested() {
        return window.location.pathname.split('/')[window.location.pathname.split('/').length - 1];
    }

    render() {

        return <MainPage articles={this.state.articles} issue={this.state.issue} />;
    }

}

const MainPageContainerWithData = compose((
    graphql(ArticlePreviewTagQuery, {name: 'articleTagPreviews'}),
    graphql(ArticlePreviewIssueQuery, {name: 'articleIssuePreviews'}),
    graphql(IssueInfoQuery, {name: 'issueInfo'})
))(MainPageContainer);

export default withApollo(MainPageContainerWithData);
