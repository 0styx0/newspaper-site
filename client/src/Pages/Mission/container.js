import React from 'react';

import httpNotification from '../../helpers/Notification';
import fetchFromApi from '../../helpers/fetchFromApi';

import Mission from './';

export default class MissionContainer extends React.Component {

    constructor() {
        super();

        this.submit = this.submit.bind(this);

        this.state = {
            content: ""
        }
    }

    async componentWillMount() {

        const mission = await fetch("./missionView.html")
                                .then(data => data.text());
        this.setState({
            content: mission
        });
    }

    submit() {


        fetchFromApi("mission", "put", {
                edit: this.state.content
        })
        .then((response) => {

            httpNotification(response.status, response.statusText);
        });

    }

    render() {

        return <Mission
                 content={this.state.content}
                 submit={this.submit}
                 save={(e) => this.setState({content: e.target.innerHTML})}
               />
    }
}