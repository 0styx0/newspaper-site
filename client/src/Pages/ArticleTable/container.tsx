import * as React from 'react';
import { ArticleQuery, ArticleUpdate, ArticleDelete } from '../../graphql/articles';
import { compose, graphql, withApollo } from 'react-apollo';

import ArticleTable from './';

export interface Article {
    tags: {
            all: string[];
        };
    url: string;
    id: string;
    displayOrder: number;
    dateCreated: string;
    views: number;
    author: {
        fullName: string;
        profileLink: string;
    };
}

export interface Issue {
    num: number;
    max: number;
}

interface Props {
    data: {
        loading: boolean;
        issues?: [
            Issue & {
                articles: Article[]
            }
        ]; // will never be more length than 1
    };
    client: {
        query: ( params: { query: typeof ArticleQuery, variables: { issue: number | null; } } ) => Promise<Props>;
    };
    updateArticle: Function;
    deleteArticle: Function;
}

interface State {
    issue: Issue;
    articles: Article[];
    updates: {
        idsToDelete: Set<string>;
        displayOrder: Map<string, number>; // id => info
        tags: Map<string, string[]>;
    };
}

/**
 * Controller of ArticleTable (@see ./index)
 *
 * Gets data from server, then renders ArticleTable with it.
 * This holds listeners for all events that @see ArticleTable can fire
 */
export class ArticleTableContainer extends React.Component<Props, State> {

    constructor() {
        super();

        this.putData = this.putData.bind(this);
        this.convertPropsToState = this.convertPropsToState.bind(this);
        this.onChange = this.onChange.bind(this);
        this.onDelete = this.onDelete.bind(this);
        this.onSubmit = this.onSubmit.bind(this);



        this.state = {
            issue: {} as Issue,
            articles: [],
            updates: {
                idsToDelete: new Set<string>(),
                displayOrder: new Map<string, number>(),
                tags: new Map<string, string[]>()
            }
        };
    }

    /**
     * @param num - articles from what issue to get. If null, get latest
     *
     * Gets data needed for table (@see State.Article interface) and changes page url to modifyArticles/currentIssue
     *
     * Depends on window.location.pathname
     */
    async putData(num: number | null = null) {

        const lastPath = +window.location.pathname.split('/').slice(-1)[0];

        num = (isNaN(lastPath) || ((!isNaN(+num!)) && num)) ? num : lastPath;

        this.props.client.query({
            query: ArticleQuery,
            variables: {
                issue: num
            }
        }).then((data: Props) => {
            this.convertPropsToState(data);
        });

        window.history.pushState(
            {},
            `Issue ${num}`, isNaN(lastPath) ?
              `modifyArticles/${num}` :
             num + ''
        );
    }

    /**
     * Seeds state with graphql props
     */
    componentWillReceiveProps(newProps: Props) {
        this.convertPropsToState(newProps);
    }

    /**
     * Sets `state.issue` and `state.articles` to values extracted from props
     */
    convertPropsToState(props: Props) {

        if (!props.data.issues) {
            return;
        }

        let {max, num} = props.data.issues![0];

        this.setState({
            issue: {
                max,
                num
            },
            articles: props.data.issues![0].articles
        });
    }

    /**
     * Saves changes to state so can be submitted later on. This is for updates only. @see onDelete for deletions
     *
     * @uses `e.target.name` as index of state.updates
     * @uses `e.target.value` as the value
     */
    onChange(e: Event, article: Article) {

        e.stopPropagation();
        e.preventDefault();

        const target = e.target as HTMLInputElement | HTMLSelectElement;

        const stateUpdate = this.state.updates;

        const modifyMapCopy = new Map(stateUpdate[target.name]);
        let value: number | string[];

        if ((target as HTMLSelectElement).selectedOptions) { // for tags

            value = Array.from((target as HTMLSelectElement).selectedOptions)
                    .map(option => option.value)
                    .slice(-3)
                    .sort(); // using sort to make it easier to test, no actual reason besides that
        } else {
            value = +target.value;
        }

        modifyMapCopy.set(article.id, value);

        stateUpdate[target.name] = modifyMapCopy;

        this.setState({
            updates: stateUpdate
        });
    }

    /**
     * Adds/removes user's id from list of users to delete
     *
     * @uses e.target.checked as way to see if should add or remove e.target.value from list
     *
     * @param e {HTMLCheckboxEvent}
     */
    onDelete(e: Event) {

        const stateUpdate = this.state.updates;

        const target = e.target as HTMLInputElement;

        target.checked ? stateUpdate.idsToDelete.add(target.value) : stateUpdate.idsToDelete.delete(target.value);

        this.setState({
            updates: stateUpdate
        });
    }

    /**
     * Sends data from `state.updates` to server
     */
    onSubmit(e: Event) {

        e.stopPropagation();
        e.preventDefault();

        // separating update and delete into functions just to show they're separate stuff
        const submitUpdated = () => {

            const { displayOrder, tags } = this.state.updates;

            if (displayOrder.size + tags.size < 1) {
                return;
            }

            let data: {id: string, displayOrder?: number, tags?: string[]}[] = [...displayOrder].map(mapping => ({

                id: mapping[0],
                displayOrder: mapping[1]
            }));

            [...tags].forEach(mapping => {

                const idIdx = data.findIndex(elt => elt.id === mapping[0]);

                if (idIdx !== -1) {
                    data[idIdx].tags = mapping[1];
                } else {
                    data.push({
                        id: mapping[0],
                        tags: mapping[1]
                    });
                }
            });

            this.props.updateArticle({
                variables: {
                    data
                }
            });
        };

        const submitDeleted = () => {

            if (this.state.updates.idsToDelete.size > 0) {

                this.props.deleteArticle({
                    variables: {
                        ids: [...this.state.updates.idsToDelete]
                    }
                });
            }
        };

        submitUpdated();
        submitDeleted();
    }

    render() {

        if (this.state.articles.length < 1) {
            return null;
        }

        return (
            <ArticleTable
                articles={this.state.articles}
                key={this.state.issue.num}
                issue={this.state.issue}
                onChange={this.onChange}
                onDelete={this.onDelete}
                onUpdate={(e: Event) => this.putData(+(e.target as HTMLInputElement).value)}
                onSubmit={this.onSubmit}
            />
        );
    }
}

const ArticleTableContainerWithData = compose(
    graphql(ArticleQuery, {
        options: {
            variables: {
                issue: +window.location.pathname.split('/').slice(-1)[0]
            }
        }
    }),
    graphql(ArticleUpdate, {name: 'updateArticle'}),
    graphql(ArticleDelete, {name: 'deleteArticle'})
)(ArticleTableContainer as any);

export default withApollo(ArticleTableContainerWithData);