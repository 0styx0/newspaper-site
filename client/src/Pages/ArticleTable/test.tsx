import * as React from 'react';
import { ArticleTableContainer, Article, Issue } from './container';
import { mount } from 'enzyme';
import * as renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router';
import localStorageMock from '../../tests/localstorage.mock';
import * as casual from 'casual';

localStorageMock.setItem('jwt', JSON.stringify([,{level: 3}])); // only lvl 3 can access this page

casual.define('articles', function(amount: number, issue: number) {

    let articles: (Issue & { articles: Article[] })[] = [
        {
            num: issue || casual.integer(1, 100),
            max: casual.integer(1, 100),
            articles: []
        }
    ];

    while (amount-- > 0) {

        articles[0].articles.push({
            tags: {
                all: casual.word
            },
            url: encodeURIComponent(casual.title),
            id: casual.word,
            displayOrder: casual.integer(0, 100),
            dateCreated: casual.date('YYYY-MM-DD'),
            views: casual.integer(0, 100),
            author: {
                fullName: casual.first_name,
                profileLink: casual.word
            }
        });
    }

    return articles;
});

casual.define('data', (issue?: number) => ({
    loading: false,
    issues: casual.articles(casual.integer(1, 10), issue)
}));

function setup(mockGraphql: {updateArticle?: Function, deleteArticle?: Function} = {}) {

    const filler = () => true;

    return mount(
        <MemoryRouter>
            <ArticleTableContainer
                data={casual.data}
                client={{
                    query: async (query: {variables: {issue: number}}) => (
                      {data: casual.data(query.variables.issue)}
                    )
                }}
                updateArticle={mockGraphql.updateArticle ? mockGraphql.updateArticle : filler}
                deleteArticle={mockGraphql.deleteArticle ? mockGraphql.deleteArticle : filler}
            />
        </MemoryRouter>
    );
}

describe('<ArticleTableContainer>', () => {

    let wrapper: any;

    beforeEach(() => {
        wrapper = setup();
    });

    describe('snapshots', () => {

        it('renders correctly', () => {

            const tree = renderer.create(

                <ArticleTableContainer
                    data={casual.data}
                    updateArticle={() => true}
                    deleteArticle={() => true}
                    client={{
                        query: () => true
                    }}
                />
            ).toJSON();

            expect(tree).toMatchSnapshot();
        });
    });

});
