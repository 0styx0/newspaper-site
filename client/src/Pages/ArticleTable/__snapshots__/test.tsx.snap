// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleTableContainer> snapshots renders correctly 1`] = `null`;
